From db2fc43169ac46152b106f71478930b5c9fa378b Mon Sep 17 00:00:00 2001
From: Sowmiya Sree Elavalagan <quic_ssreeela@quicinc.com>
Date: Sun, 15 Dec 2024 20:45:49 +0530
Subject: [PATCH] UPSTREAM:PLATFORM:wifi:ath12k:Fix crash in FW stats free path

When MLO setup fails, in error path we try to unregister and free
debugfs and FW stats, but these are initialized after MLO setup in
MAC register. Hence trying to free uninitialized member causes crash.
Fix this by freeing only if it initialized.

Signed-off-by: Sowmiya Sree Elavalagan <quic_ssreeela@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 12 +++++++++++-
 drivers/net/wireless/ath/ath12k/dp.c   |  2 --
 2 files changed, 11 insertions(+), 3 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -1812,7 +1812,8 @@ err_setup:
 static int ath12k_core_hw_group_start(struct ath12k_hw_group *ag)
 {
 	struct ath12k_base *ab = ag->ab[0];
-	int ret, i;
+	int ret, i, j;
+	struct ath12k *ar;
 	bool is_registered = false;
 
 	lockdep_assert_held(&ag->mutex_lock);
@@ -1854,7 +1855,7 @@ static int ath12k_core_hw_group_start(st
 	ret = ath12k_core_mlo_setup(ag);
 	if (ret) {
 		ath12k_err(NULL, "failed to setup MLO: %d\n", ret);
-		goto err_pdev_destroy;
+		goto err_pdev_destroy_all;
 	}
 
 	if (!is_registered) {
@@ -1944,12 +1945,29 @@ pdev_cleanup:
 err_mlo_teardown:
 	ath12k_core_mlo_teardown(ag);
 	i = ag->num_chip;
+	for (i = i - 1; i >= 0; i--) {
+		ab = ag->ab[i];
+		for (j = 0; j < ab->num_radios; j++) {
+			ar = ab->pdevs[j].ar;
+			ath12k_fw_stats_free(&ar->fw_stats);
+			ath12k_debugfs_unregister(ar);
+		}
+	}
+err_pdev_destroy_all:
+	i = ag->num_chip;
 err_pdev_destroy:
 	for (i = i - 1; i >= 0; i--) {
 		ab = ag->ab[i];
 
 		mutex_lock(&ab->core_lock);
-		ath12k_core_pdev_destroy(ab);
+		if (!ab->hw_params->rxdma1_enable)
+			del_timer_sync(&ab->mon_reap_timer);
+
+		for (j = 0; j < ab->num_radios; j++) {
+			ath12k_dp_rx_pdev_free(ab, j);
+			ath12k_dp_rx_pdev_mon_detach(ab, j);
+		}
+		ath12k_debugfs_pdev_destroy(ab);
 		mutex_unlock(&ab->core_lock);
 	}
 
